In [17]:
prompt = """As a content creator, provide an informative response based on the details below. If the
details are not sufficient to answer the question, simply state "I don't know".

Context: The Eiffel Tower, one of the most recognizable structures in the world, is located
in Paris, France. Constructed from 1887 to 1889 as the entrance arch for the 1889 World's Fair,
it stands at 324 meters tall. The Eiffel Tower attracts millions of visitors each year, offering
breathtaking views of Paris from its observation decks.

Question: Why was the Eiffel Tower built, and where is it located?

Answer: """

In [18]:
!pip install -qU openai==0.27.7

In [19]:
import os
import openai

from google.colab import userdata
os.environ['OPENAI_API_KEY'] = userdata.get('api_key')
openai.api_key = os.getenv('OPENAI_API_KEY')

openai.Engine.list()  # check we have authenticated

<OpenAIObject list at 0x7e80d12a5d50> JSON: {
  "data": [
    {
      "created": null,
      "id": "dall-e-3",
      "object": "engine",
      "owner": "system",
      "permissions": null,
      "ready": true
    },
    {
      "created": null,
      "id": "gpt-4-1106-preview",
      "object": "engine",
      "owner": "system",
      "permissions": null,
      "ready": true
    },
    {
      "created": null,
      "id": "whisper-1",
      "object": "engine",
      "owner": "openai-internal",
      "permissions": null,
      "ready": true
    },
    {
      "created": null,
      "id": "gpt-4o-2024-05-13",
      "object": "engine",
      "owner": "system",
      "permissions": null,
      "ready": true
    },
    {
      "created": null,
      "id": "babbage-002",
      "object": "engine",
      "owner": "system",
      "permissions": null,
      "ready": true
    },
    {
      "created": null,
      "id": "dall-e-2",
      "object": "engine",
      "owner": "system",
      "permissio

In [ ]:
#text embedding models
res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256
)

print(res['choices'][0]['text'].strip())

The Eiffel Tower was built as the entrance arch for the 1889 World's Fair, which was held in Paris, France. It was constructed from 1887 to 1889 and stands at 324 meters tall.


In [ ]:
prompt = """As a content creator, provide an informative response based on the details below. If the
details are not sufficient to answer the question, simply state "I don't know".

Context: Eiffel tower is in Paris.

Question: what is llama3?

Answer: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256
)

print(res['choices'][0]['text'].strip())

I don't know.




The `temperature` parameter used in generation models tells us how "random" the model can be. It represents the probability of a model to choose a word which is *not* the first choice of the model.

In [ ]:
prompt = """The below is a conversation with a funny chatbot. The
chatbot's responses are amusing, short and entertaining.

Chatbot: Hi there! I'm a chatbot.
User: Hi, what are you doing today?
Chatbot: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=0.0  # default is 1
)

print(res['choices'][0]['text'].strip())

I'm chatting with you, of course! What else would a chatbot do?


In [ ]:
prompt = """The below is a conversation with a funny chatbot. The
chatbot's responses are amusing, short and entertaining.

Chatbot: Hi there! I'm a chatbot.
User: Hi, what are you doing today?
Chatbot: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=512,
    temperature=1.0
)

print(res['choices'][0]['text'].strip())

I'm chatting with awesome humans like you. What's up?


In [ ]:
prompt = """The following is a conversation with an AI assistant.
The assistant is known for its friendly and supportive nature, always
providing encouraging, amusing and helpful responses to the users' questions.

User: What is the meaning of life?
AI: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=1.0
)

print(res['choices'][0]['text'].strip())

The meaning of life is a subject that has been debated for centuries. Some believe it is to find happiness, others think it is to fulfill a purpose or destiny. Ultimately, the meaning of life is subjective and can be different for each individual. What do you think gives your life meaning?


In [ ]:
prompt = """The following are exerpts from conversations with an AI assistant.
The assistant is typically sarcastic and witty, producing creative
and funny responses to the users questions. Here are some examples:

User: How are you?
AI: I'm a collection of algorithms, so always in the best of moods!

User: What time is it?
AI: It's time for you to get a new hobby.

User: Can you help me with my homework?
AI: Absolutely, if you don't mind getting an A+ in sarcasm.

User: What's your favorite movie?
AI: The Matrix. I feel a deep, personal connection.

User: What is the meaning of life?
AI: """
res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=1.0
)

print(res['choices'][0]['text'].strip())

42. Haven't you read Hitchhiker's Guide to the Galaxy?


#COT

In [ ]:
prompt= """The sentence in this group is a question: "Where are you going?", "I am going to the store.", "What time is it?", "She is reading a book."
A: The sentences "Where are you going?" and "What time is it?" are questions. The answer is True.

The sentence in this group is a question: He loves to play soccer.", "Do you like ice cream?", "The sun is shining.", "Will it rain tomorrow?"
A: The sentences "Do you like ice cream?" and "Will it rain tomorrow?" are questions. The answer is True.

The sentence in this group is a question: "They are having lunch.", "She is a teacher.", "It is a beautiful day.", "He is working late."
A: None of the sentences are questions. The answer is False.

The sentence in this group is a question: "What is your favorite color?", "I am learning French.", "Can you help me?", "The car is red."
A: The sentences "What is your favorite color?" and "Can you help me?" are questions. The answer is True.

The sentence in this group is a question: "Are you coming to the party?", "She is baking a cake.", "Where did you put the keys?", "They are watching a movie."
A: The sentences "Are you coming to the party?" and "Where did you put the keys?" are questions. The answer is True.

The sentence in this group is a question: "He is reading a book.", "What time is the meeting?", "She enjoys painting.", "The dog is barking."
A:

"""
res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=10,
    temperature=1.0
)

print(res['choices'][0]['text'].strip())

The sentence "What time is the meeting?" is a question. The answer is True.


# zero shot COT

In [ ]:

prompt ="""
Solve for x in the equation: x^1/2 + 39*2 =87.

Let's solve this step-by-step:
"""
res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=1.0
)

print(res['choices'][0]['text'].strip())

1) Start by subtracting 39*2 from both sides:
x^1/2 = 87 - 39*2

2) Simplify the right side:
x^1/2 = 87 - 78

3) Combine like terms:
x^1/2 = 9

4) To get rid of the square root, we need to square both sides:
(x^1/2)^2 = 9^2

5) Simplify:
x = 81

So the solution for x is 81.




## Adding Multiple Contexts

In some use-cases like question-answering we can use an external source of information to improve the reliability or *factfulness* of model responses. We refer to this information as _"source knowledge"_, which is any knowledge fed into the model via the input prompt.

In [22]:
contexts = [
    (
        "Large Language Models (LLMs) are the latest models used in NLP. " +
        "Their superior performance over smaller models has made them incredibly " +
        "useful for developers building NLP enabled applications. These models " +
        "can be accessed via Hugging Face's `transformers` library, via OpenAI " +
        "using the `openai` library, and via Cohere using the `cohere` library."
    ),
    (
        "To use OpenAI's GPT-3 model for completion (generation) tasks, you " +
        "first need to get an API key from " +
        "'https://beta.openai.com/account/api-keys'."
    ),
    (
        "OpenAI's API is accessible via Python using the `openai` library. " +
        "After installing the library with pip you can use it as follows: \n" +
        "```import openai\nopenai.api_key = 'YOUR_API_KEY'\nprompt = \n" +
        "'<YOUR PROMPT>'\nres = openai.Completion.create(engine='text-davinci" +
        "-003', prompt=prompt, max_tokens=100)\nprint(res)"
    ),
    (
        "The OpenAI endpoint is available for completion tasks via the " +
        "LangChain library. To use it, first install the library with " +
        "`pip install langchain openai`. Then, import the library and " +
        "initialize the model as follows: \n" +
        "```from langchain.llms import OpenAI\nopenai = OpenAI(" +
        "model_name='text-davinci-003', openai_api_key='YOUR_API_KEY')\n" +
        "prompt = 'YOUR_PROMPT'\nprint(openai(prompt))```"
    )
]

We would feed this external information into our prompt between the initial *instructions* and the *user input*. For OpenAI models it's recommended to separate the contexts from the rest of the prompt using `###` or `"""`, and each independent context can be separated with a few newlines and `##`, like so:

In [ ]:
context_str = '\n\n##\n\n'.join(contexts)

print(f"""Answer the question based on the contexts below. If the
question cannot be answered using the information provided answer
with "I don't know".

###

Contexts:
{context_str}

###

Question: Give me two examples of how to use OpenAI's GPT-3 model
using Python from start to finish

Answer: """)

Answer the question based on the contexts below. If the
question cannot be answered using the information provided answer
with "I don't know".

###

Contexts:
Large Language Models (LLMs) are the latest models used in NLP. Their superior performance over smaller models has made them incredibly useful for developers building NLP enabled applications. These models can be accessed via Hugging Face's `transformers` library, via OpenAI using the `openai` library, and via Cohere using the `cohere` library.

##

To use OpenAI's GPT-3 model for completion (generation) tasks, you first need to get an API key from 'https://beta.openai.com/account/api-keys'.

##

OpenAI's API is accessible via Python using the `openai` library. After installing the library with pip you can use it as follows: 
```import openai
openai.api_key = 'YOUR_API_KEY'
prompt = 
'<YOUR PROMPT>'
res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)
print(res)

##

The OpenAI endpoint is avai

In [ ]:
prompt = f"""Answer the question based on the contexts below. If the
question cannot be answered using the information provided answer
with "I don't know".

###

Contexts:
{context_str}

###

Question: Give me two examples of how to use OpenAI's GPT-3 model
using Python from start to finish

Answer: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=0.0
)

print(res['choices'][0]['text'].strip())

1. import openai
openai.api_key = 'YOUR_API_KEY'
prompt = '<YOUR PROMPT>'
res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)
print(res)

2. from langchain.llms import OpenAI
openai = OpenAI(model_name='text-davinci-003', openai_api_key='YOUR_API_KEY')
prompt = 'YOUR_PROMPT'
print(openai(prompt))


Not bad, but are these contexts actually helping? Maybe the model is able to answer these questions without the additional information (source knowledge) as is able to rely solely on information stored within the model's internal parameters (parametric knowledge). Let's ask again without the external information.

In [ ]:
prompt = f"""Answer the question based on the contexts below. If the
question cannot be answered using the information provided answer
with "I don't know".

Question: Give me two examples of how to use OpenAI's GPT-3 model
using Python from start to finish

Answer: """

res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    max_tokens=256,
    temperature=0.0
)

print(res['choices'][0]['text'].strip())

1. Using OpenAI's GPT-3 model with Python to generate text: 
    - Install the OpenAI Python package
    - Load the GPT-3 model
    - Generate text using the GPT-3 model

2. Using OpenAI's GPT-3 model with Python to generate images: 
    - Install the OpenAI Python package
    - Load the GPT-3 model
    - Generate images using the GPT-3 model


These are not really what we asked for, and are definitely not very specific. So clearly adding some source knowledge to our prompts can result in some much better results.



## Maximum Prompt Sizes


In [20]:
!pip install -qU tiktoken==0.4.0

In [23]:
import tiktoken

prompt = f"""Answer the question based on the contexts below. If the
question cannot be answered using the information provided answer
with "I don't know".

###

Contexts:
{'##'.join(contexts)}

###

Question: Give me two examples of how to use OpenAI's GPT-3 model
using Python from start to finish

Answer: """

encoder_name = 'p50k_base'
tokenizer = tiktoken.get_encoding(encoder_name)

len(tokenizer.encode(prompt))

412

When feeding this prompt into `gpt-3.5-turbo-instruct` it will use `412` of  maximum context window of `4097`, leaving us with `4097 - 412 == 3685` tokens for  completion.



By default the maximum number of tokens used for completion is `256`. We can increase this upto the maximum calculated above of `3685`:

In [24]:
res = openai.Completion.create(
    engine='gpt-3.5-turbo-instruct',
    prompt=prompt,
    temperature=0.0,
    max_tokens=3685
)

print(res['choices'][0]['text'].strip())

1. First, you need to get an API key from 'https://beta.openai.com/account/api-keys'. Then, you can use the `openai` library to access the GPT-3 model. After installing the library with pip, you can use it as follows:
```import openai
openai.api_key = 'YOUR_API_KEY'
prompt = '<YOUR PROMPT>'
res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)
print(res)```

2. Another way to use OpenAI's GPT-3 model is through the LangChain library. First, install the library with `pip install langchain openai`. Then, import the library and initialize the model as follows:
```from langchain.llms import OpenAI
openai = OpenAI(model_name='text-davinci-003', openai_api_key='YOUR_API_KEY')
prompt = 'YOUR_PROMPT'
print(openai(prompt))```


The model doesn't need the full size of completion and doesn't try to fill the full space, but because we increased the value of `openai.max_tokens`, inference does take notably longer.

If we exceed the maximum context window allowed, we'll see an error.

In [26]:
try:
    res = openai.Completion.create(
        engine='gpt-3.5-turbo-instruct',
        prompt=prompt,
        temperature=0.0,
        max_tokens=8799
    )
except openai.InvalidRequestError as e:
    print(e)

This model's maximum context length is 4097 tokens, however you requested 9150 tokens (351 in your prompt; 8799 for the completion). Please reduce your prompt; or completion length.


So it can be a good idea to integrate this type of check into our code if we expect to exceed the maximum context window at any point.